### PR TITLE
Changed default memory layout of MemoryReader to 'fac'

### LINKED ISSUE #1063: Default numpy array layout for MemoryReader
# MemoryReader layout
1. Change the MemoryReader default order ("format") to fac (to be compatible with `AnalysisFromFunction()` and friends.
2. Change the keyword argument "format" to "order" (and check other parts of the code (eg ENCORE) as well as the docs to follow this change)

- [ ] changed MemoryReader
- [ ] changed docs (see #1056)
- [ ] tests

# Background
There are currently two different default layouts for numpy arrays in MDAnalysis. `DCDReader.timeseries()` uses an 'afc' format by default, which has now also been adopted as the default in `MemoryReader`, both when constructing a new `MemoryReader`, and when extracting the raw array using `MemoryReader.timeseries()`.
On the other hand, the natural order when using `AnalysisFromFunction` is 'fac', which also intuitively (at least to me) is a more natural order.

These different default layouts leads to some unnecessary confusion. Note, for instance, the `swapaxes` that is necessary when extracting coordinates using `AnalysisFromFunction` to use to construct a new Universe with MemoryReader:
```
  coordinates = AnalysisFromFunction(lambda ag: ag.positions.copy(),
                                     u.atoms).run().results.swapaxes(0, 1)
  u2 = mda.Universe(PDB, coordinates, format=MemoryReader)
```

Since `MemoryReader` is introduced in this release, it is still relatively easy to change the default layout of MemoryReader, without having to worry about breaking code. So, the question is: should we change the default layout of MemoryReader so that it upon construction uses 'fac' instead? 
And if so, do we then keep the default of `MemoryReader.timeseries()` to be 'afc', to maintain consistency with `DCDReader.timeseries`?

### Currently version of MDAnalysis:
0.16.0-dev0

# History
* 2016-11-03 edited by @orbeckst (added agreed upon course of action and check list)
* 2016-11-06 edited by @orbeckst (change "format" to "order")

### PR DESCRIPTION
**Pull request by wouterboomsma**
Fixes #1063 and #1065

Changes made in this Pull Request:
 - Switched default layout of MemoryReader to 'fac'
 - Ensure that the internal numpy array in a MemoryReader is always float32.

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by wouterboomsma** [comment]
Sorry for the delay.

One small annoyance is that this code no longer works

```python
    coordinates = universe.trajectory.timeseries(universe.atoms)
    universe2 = mda.Universe(PSF, coordinates, format=MemoryReader)
```
and has to be replaced with
```python
    coordinates = universe.trajectory.timeseries(universe.atoms, format='fac')
    universe2 = mda.Universe(PSF, coordinates, format=MemoryReader)
```
or
```python
    coordinates = universe.trajectory.timeseries(universe.atoms)
    universe2 = mda.Universe(PSF, coordinates, format=MemoryReader, order='afc')
```
Encore uses `timeseries()` interface quite frequently to extract subuniverses with `MemoryReader` so this means I had to add `format='fac'` quite a few places. But I can live with this solution.

**Comment by mnmelo** [comment]
Perhaps this makes the case for a change of the timeseries return default to `'fac'`?
I'd be for it, since I find fac order much more intuitive.

**Comment by wouterboomsma** [comment]
@mnmelo I agree about the intuition, although I'm a bit worried about how much user-code we would break by changing the default of timeseries. But at least changing the MemoryReader to the right layout now before release is a step in the right direction.

File: package/MDAnalysis/analysis/encore/bootstrap.py [hunk 1]
```diff
@@ -150,6 +150,6 @@ def get_ensemble_bootstrap_samples(ensemble,
             size=ensemble.trajectory.timeseries().shape[1])
         ensembles.append(
             mda.Universe(ensemble.filename,
-                        ensemble.trajectory.timeseries(format='afc')[:,indices,:],
+                        ensemble.trajectory.timeseries(format='fac')[indices,:,:],
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/encore/bootstrap.py]: you can now actually skip the last `:, :]` in the fancy index. Numpy will default to taking the full array for lower dimensions

File: package/MDAnalysis/coordinates/memory.py [hunk 2]
```diff
@@ -167,7 +161,7 @@
 
   u2 = mda.Merge(protein).load_new(
            AnalysisFromFunction(lambda ag: ag.positions.copy(),
-                                protein).run().results.swapaxes(0, 1),
```
Comment (kain88-de) [inline package/MDAnalysis/coordinates/memory.py]: It's really nice that the swapaxes are gone now. That makes the code easier to understand

**Comment by kain88-de** [comment]
@wouterboomsma ping

**Comment by wouterboomsma** [comment]
@kain88-de, @orbeckst  I'm back. Sorry for the delay. I'll implement the changes requested by @orbeckst today or tomorrow. Two comments:

I sometimes add the extra :,: to make it clear to the reader of the code how many dimensions the container has - which is not always clear from the context. But I can certainly remove this if you like.

@orbeckst, when you write "test that checks the default memory order" do you mean tests that verify that after reading in a structure from file, the layout in the Memory Reader corresponds to the 'fac' standard. So for instance, reading it in with and with and without explicitly setting the format to 'fac' and then comparing that they are the same?

**Comment by orbeckst** [comment]
Yes!


> Am Jan 12, 2017 um 7:40 schrieb Wouter Boomsma <notifications@github.com>:
> 
> @orbeckst, when you write "test that checks the default memory order" do you mean tests that verify that after reading in a structure from file, the layout in the Memory Reader corresponds to the 'fac' standard. So for instance, reading it in with and with and without explicitly setting the format to 'fac' and then comparing that they are the same?

**Comment by jbarnoud** [comment]
> I sometimes add the extra :,: to make it clear to the reader of the code how many dimensions the container has - which is not always clear from the context. But I can certainly remove this if you like.

For what it is worth, I prefer the explicit notation.

**Comment by wouterboomsma** [comment]
@orbeckst Added test for default layout (2be6477)

**Comment by orbeckst** [comment]
Thanks. Will look later today. Really busy day. 

--
Oliver Beckstein
email: orbeckst@gmail.com

**Comment by orbeckst** [comment]
All good, thank you. Can this be merged from your end?

**Comment by wouterboomsma** [comment]
@orbeckst I don't think I have permissions to merge anything into the main repo (at least, I cannot see any "merge" button - could be due to my limited experience with github).

**Comment by kain88-de** [comment]
No only maintainers can merge you haven't overlooked anything.

**Pull request merged**